pkg/offline/offline.go

```diff
@@ -157,6 +167,164 @@ func WithQueue(filepath string, syncLimit int) (heartbeat.HandleOption, error) {
 	}, nil
 }
 
+// Sync returns a function to send queued heartbeats to the WakaTime API.
+func Sync(filepath string, syncLimit int) func(next heartbeat.Handle) error {
+	return func(next heartbeat.Handle) error {
+		log.Debugf("execute offline sync with file %s", filepath)
+
+		hh, err := popHeartbeats(filepath, sendLimit)
+		if err != nil {
+			return fmt.Errorf("failed to fetch heartbeat from offline queue: %s", err)
+		}
+
+		if len(hh) == 0 {
+			log.Debugln("abort execution, as there are no queued heartbeats ready for sending")
+
+			return nil
+		}
+
+		results, err := next(hh)
+		if err != nil {
+			var count int
+
+			for {
+				if count >= maxRequeueAttempts {
+					log.Errorf("abort requeuing after %d unsuccessful attempts", count)
+					break
+				}
+
+				requeueErr := pushHeartbeats(filepath, hh)
+				if requeueErr != nil {
+					log.Errorf("failed to push heatbeats after api error to queue: %s", requeueErr)
```

Comment:
This should be a warning.

Reply:
I guess we're best here with warning. I mean something really went wrong here, and also if we want to recover lost heartbeats, we don't want to depend on the user setting log level to debug, right?

Reply:
We've already got the error logging at `if count >= maxRequeueAttempts`, so we don't need to log two errors for the same thing. Here we can still retry, so the heartbeat might not be lost. If we retry, nothing should be printed unless `--verbose` was passed. That's because it's highly possible two wakatime-cli processes try to write at the same time, with one not able to acquire the global lock. That's not a problem, as long as we don't reach the `maxRequeueAttempts`.

Reply:
Yeah, this is a good idea. Adjusted following your remarks :+1: